### PR TITLE
A re-open PR: Avoid re-creating the random number generator in RandomSampler

### PR DESCRIPTION
**Pull request by lartpang**
More details can be found in the old pr: https://github.com/pytorch/pytorch/pull/53085

@ejguan  Thanks for your guidance. I tried to reopen this PR following your instructions.

### CODE REVIEW

**Review by ejguan — APPROVED** [review]

**Comment by vfdev-5** [comment]
@ejguan @lartpang I think this PR can be BC-breaking. For example:
```python
import torch
print(torch.__version__)

from torch.utils.data import DataLoader

data = range(20)

print("--- set random seed")
torch.manual_seed(1)
loader = DataLoader(data, batch_size=4, shuffle=True)
for b in loader:
    print(b)

print("--- set random seed")
torch.manual_seed(1)
for b in loader:
    print(b)
```
Output for v1.9.0
```
1.9.0
--- set random seed
tensor([14, 13, 12,  8])
tensor([ 3, 17,  7,  6])
tensor([10,  1,  4,  0])
tensor([19, 18, 16, 15])
tensor([ 5,  9, 11,  2])
--- set random seed
tensor([14, 13, 12,  8])
tensor([ 3, 17,  7,  6])
tensor([10,  1,  4,  0])
tensor([19, 18, 16, 15])
tensor([ 5,  9, 11,  2])
```
and nightly with this PR:
```
1.10.0.dev20210819+cu111
--- set random seed
tensor([14, 13, 12,  8])
tensor([ 3, 17,  7,  6])
tensor([10,  1,  4,  0])
tensor([19, 18, 16, 15])
tensor([ 5,  9, 11,  2])
--- set random seed
tensor([8, 7, 6, 3])
tensor([13, 12, 15,  1])
tensor([ 9, 14, 16, 11])
tensor([18,  2, 19, 10])
tensor([ 4, 17,  5,  0])
```

This may impact users who would like to control the randomness for reproducibility. 
What do you think ?

**Comment by ejguan** [comment]
@vfdev-5 Thanks for pointing it out. But, it actually doesn't break the reproducibility. IMO, reproducibility refers to the data flow can be reproduced with same code. If you set seed before DataLoader is constructed, the data flow is deterministic. And, we do not guarantee the randomness across PyTorch release. https://pytorch.org/docs/stable/notes/randomness.html

But, I agree with you that users would expect to yield same data after setting seed. Just wondering is this the normal use case for training? I can send a PR to fix it rather reverting this PR, since it does bring on the benefit to serialize the state of generator during training.

Since I am also working on determinism for the new DataLoader/DataPipe system, I would appreciate if you can provide some insights on the deterministic data flow.
- How important is deterministic data pipeline? And, if there is performance downgrade, do you still want the whole pipeline deterministic?
  - WDYT about an option for users to choose between determinism and performance?
- What's your use case when you want to enable the deterministic data flow?

**Comment by lartpang** [comment]
> @ejguan @lartpang I think this PR can be BC-breaking. For example:
> 
> ```python
> import torch
> print(torch.__version__)
> 
> from torch.utils.data import DataLoader
> 
> data = range(20)
> 
> print("--- set random seed")
> torch.manual_seed(1)
> loader = DataLoader(data, batch_size=4, shuffle=True)
> for b in loader:
>     print(b)
> 
> print("--- set random seed")
> torch.manual_seed(1)
> for b in loader:
>     print(b)
> ```
> 
> Output for v1.9.0
> 
> ```
> 1.9.0
> --- set random seed
> tensor([14, 13, 12,  8])
> tensor([ 3, 17,  7,  6])
> tensor([10,  1,  4,  0])
> tensor([19, 18, 16, 15])
> tensor([ 5,  9, 11,  2])
> --- set random seed
> tensor([14, 13, 12,  8])
> tensor([ 3, 17,  7,  6])
> tensor([10,  1,  4,  0])
> tensor([19, 18, 16, 15])
> tensor([ 5,  9, 11,  2])
> ```
> 
> and nightly with this PR:
> 
> ```
> 1.10.0.dev20210819+cu111
> --- set random seed
> tensor([14, 13, 12,  8])
> tensor([ 3, 17,  7,  6])
> tensor([10,  1,  4,  0])
> tensor([19, 18, 16, 15])
> tensor([ 5,  9, 11,  2])
> --- set random seed
> tensor([8, 7, 6, 3])
> tensor([13, 12, 15,  1])
> tensor([ 9, 14, 16, 11])
> tensor([18,  2, 19, 10])
> tensor([ 4, 17,  5,  0])
> ```
> 
> This may impact users who would like to control the randomness for reproducibility.
> What do you think ?

From the results, it seems that the Generator instantiated in DataLoader is not controlled by `torch.manual_seed`.

```python
import torch
from torch import default_generator

torch.manual_seed(1)
generator = torch.Generator()
print(torch.randperm(10, generator=generator))
print(torch.randperm(10, generator=default_generator))

torch.manual_seed(1)
print(torch.randperm(10, generator=generator))
print(torch.randperm(10, generator=default_generator))
```

**Comment by vfdev-5** [comment]
Thanks for reply @ejguan ! 

> Just wondering is this the normal use case for training?

No, it is a common use case, but it can make sense when users would like to fully control the data flow. The main point of calling `torch.manual_seed` before next run of the dataloader is that during the next epoch model "sees" the same data while the first run or resuming from an epoch:
```python
dataloader = ...
for epoch in range(num_epochs):
    torch.manual_seed(seed + epoch)
    for batch in dataloader:
        ...
```

> How important is deterministic data pipeline? And, if there is performance downgrade, do you still want the whole pipeline deterministic?

From what I know, in medical applications users can choose deterministic pipeline even if there is performance downgrade.

> WDYT about an option for users to choose between determinism and performance?

Yes, it makes sense. I wonder if reusing the same generator there is a large performance speedup ?

> What's your use case when you want to enable the deterministic data flow?

One of them: resuming from a checkpoint and n-th epoch or n-th iteration (in the middle of epoch).

**Comment by ejguan** [comment]
@lartpang 
It's not a bug.
It's just different expectation for the Sampler. For the case that `Sampler` is explicitly attached a generator, it's straight forward to use a single generator instance across epoch.
And, for the case of Sample constructed without generator. As @vfdev-5 suggested, if users specify a new seed, they would expect to change the behavior of Sampler as they didn't specify the generator at the beginning. Just like all the random operators in PyTorch.
```py
>>> generator = torch.Generator()
>>> generator.manual_seed(0)
>>> torch.randn(1, generator=generator)
tensor([1.5410])
>>> torch.randn(1, generator=generator)
tensor([-0.2934])

>>> torch.manual_seed(0)
>>> torch.randn(1)
tensor([1.5410])
>>> torch.manual_seed(0)
>>> torch.randn(1)
tensor([1.5410])
```

**Comment by ejguan** [comment]
@vfdev-5 
> No, it is a common use case, but it can make sense when users would like to fully control the data flow.

Emmmm, I guess you mean it's not a common use case.

>  I wonder if reusing the same generator there is a large performance speedup ?

There won't be a huge performance speedup, IMO. And, I was referring the option of performance or determinism to the new DataLoader/DataPipe system. We want to design a system that can support both use cases.

> One of them: resuming from a checkpoint from n-th epoch or n-th iteration (in the middle of epoch).

I see. We are also working on doing checkpointing on data loading part, then you will be able to resume from it. But, tbh, I don't think previous Sampler is able to resume in the middle of epoch. You can't serialize the information of generator, IMO. Or, you are using seed as a checkpoint and start from the beginning of the epoch you just stopped, then skip feeding data to the model until reach the saved iteration number.
We want to provide a way to directly resume from the iteration step rather than start from beginning.

**Comment by vfdev-5** [comment]
> Emmmm, I guess you mean it's not a common use case.

Yes, sorry, forgot to write "not". It is not a common use case.

> And, I was referring the option of performance or determinism to the new DataLoader/DataPipe system. We want to design a system that can support both use cases.

Yes, it would be nice. Just a nit comment about a number of flags to tweak for one or another option can become a bit large, IMO.

> But, tbh, I don't think previous Sampler is able to resume in the middle of epoch. You can't serialize the information of generator, IMO. 

Yes, it is true that in general it is not straightforward. What is possible is to fetch batch indices, store them once in the beginning and then replay (we did that [here](https://github.com/pytorch/ignite/blob/5d4f86997da356ed32ca94e3d1eb8ee629525dd5/ignite/engine/deterministic.py#L38)).

> Or, you are using seed as a checkpoint and start from the beginning of the epoch you just stopped, then skip feeding data to the model until reach the saved iteration number.

yes, something like that :)

> We want to provide a way to directly resume from the iteration step rather than start from beginning.

It would be very helpful. However, a complicated thing while resuming from the iteration is to deal with data augmentations as they can be handled by a 3rd party libs...

**Comment by ejguan** [comment]
> Yes, it is true that in general it is not straightforward. What is possible is to fetch batch indices, store them once in the beginning and then replay (we did that here).

Awesome, I will take a look.

> However, a complicated thing while resuming from the iteration is to deal with data augmentations as they can be handled by a 3rd party libs...

Yep, it's kind annoying and hard to control the randomness for 3rd party libs, but it's still doable as long as 3rd party libs support random function with explicit generator, IMO. That's a great feedback, will look into it.
